nautilus_trader/backtest/config.py

```diff
@@ -0,0 +1,277 @@
+import dataclasses
+from functools import partial
+from typing import List, Optional, Tuple
+
+from dask.base import normalize_token
+from dask.base import tokenize
+from dask import delayed
+import pandas as pd
+
+from nautilus_trader.backtest.data_loader import DataCatalog
+from nautilus_trader.backtest.engine import BacktestEngine
+from nautilus_trader.backtest.models import FillModel
+from nautilus_trader.backtest.modules import SimulationModule
+from nautilus_trader.core.message import Event
+from nautilus_trader.model.c_enums.account_type import AccountTypeParser
+from nautilus_trader.model.c_enums.oms_type import OMSTypeParser
+from nautilus_trader.model.c_enums.venue_type import VenueTypeParser
+from nautilus_trader.model.currency import Currency
+from nautilus_trader.model.identifiers import Venue
+from nautilus_trader.model.instruments.base import Instrument
+from nautilus_trader.model.objects import Money
+from nautilus_trader.model.orderbook.book import OrderBookDelta
+from nautilus_trader.model.tick import QuoteTick
+from nautilus_trader.model.tick import TradeTick
+
+
+PARTIAL_SUFFIX = "Partial-"
+
+
+class Partialable:
+    def missing(self):
+        return [x for x in self.__dataclass_fields__ if getattr(self, x) is None]
+
+    def is_partial(self):
+        return any(self.missing())
+
+    def check(self, ignore=None):
+        missing = [m for m in self.missing() if m not in (ignore or {})]
+        if missing:
+            raise AssertionError(f"Missing fields: {missing}")
+
+    def _check_kwargs(self, kw):
+        for k in kw:
+            assert k in self.__dataclass_fields__, f"Unknown kwarg: {k}"
+
+    def update(self, **kwargs):
+        """Update attributes on this instance"""
+        self._check_kwargs(kwargs)
+        self.__dict__.update(kwargs)
+        return self
+
+    def replace(self, **kwargs):
+        """Return a new instance with some attributes replaces"""
+        return self.__class__(
+            **{**{k: getattr(self, k) for k in self.__dataclass_fields__}, **kwargs}
+        )
+
+    def __repr__(self):
+        dataclass_repr_func = dataclasses._repr_fn(
+            fields=list(self.__dataclass_fields__.values()), globals=self.__dict__
+        )
+        r = dataclass_repr_func(self)
+        if self.missing():
+            return "Partial-" + r
+        return r
+
+
+@dataclasses.dataclass()
```

Comment:
I don't know if you've ever used Pydantic, but for modeling/data validation, I would recommend Pydantic over standard `dataclasses`: https://pydantic-docs.helpmanual.io/

Reply:
Yeah I have used it quite a bit (it actually formed the base of the first commit - https://github.com/nautechsystems/nautilus_trader/pull/361/commits/ce31f28f898fa0eff48b0b3b68161c6538507167) but it doesn't really allow partial configuration (which I think is important for usability) and I thought it was a bit of overkill at this stage.

Reply:
Totally missed reading that! I agree, should get this working and usable first.